Add support for Jetpack products to <PurchasesListing /> on my-plan page

delawski opened the pull request:
So far the new `<PurchasesListing />` component on `/plans/my-plan` page displayed information only about the plan the site is on.

With this PR the purchases listing presents information about the Jetpack products as well.

**Please note that the `<PlanIcon />` component doesn't support Jetpack Backup yet. This will be addressed in a subsequent PR and shouldn't block this change set.**

**Before**
![Screenshot 2019-11-19 at 18 15 28](https://user-images.githubusercontent.com/478735/69169714-1ee9e180-0af9-11ea-9a85-e5c46af42214.png)

**After** 
![Screenshot 2019-11-19 at 18 14 50](https://user-images.githubusercontent.com/478735/69169717-201b0e80-0af9-11ea-82b1-8d48505ac659.png)

#### Changes proposed in this Pull Request

* Add support for Jetpack products to `<PurchasesListing />` on `/plans/my-plan` page
* Clean up `<PurchasesListing />` component
* Add Jetpack Backup tag line copy to constants file
* Add necessary utility functions to `product-values` library

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to http://calypso.localhost:3000/plans/my-plan/
* Select a Jetpack site that has a Daily backup product
* Confirm that the product is displayed in the "My Products" section
* Using SA change the expiration date to less than 30 days from now and confirm that the expiration information is now red
* Upgrade to Real-Time and confirm that the correct title and tag line is displayed

Fixes n/a

enejb commented on client/lib/products-values/index.js:
```diff
@@ -359,6 +364,25 @@ export function getDomain( product ) {
+}
+
+export function isProductExpiring( product ) {
+	const expiration = product?.expiryMoment ?? null;
```
I am not sure when this would be the case. But if we ever have a product that doesn't expire and then doesn't have a `expiryMoment` we would end up `isProductExpiring()` returning true for that case. 

Is that something we want?

tyxla replied:
I agree with Enej here.

delawski replied:
I think you're right. In this kind of an edge case we would probably want to check if we have `expiration` and if not then return `false`, i.e.:

```js
export function isProductExpiring( product ) {
	const expiration = product?.expiryMoment;

	if ( ! expiration ) {
		return false;
	}

	return expiration < moment().add( 30, 'days' );
}
```

What do you think?

enejb commented:
I left some feedback on the code which looks great! 
I don't think anything should it should hold the merge of the PR. 

But I think we should investigate this a bit.

tyxla commented on client/lib/products-values/constants.js:
```diff
@@ -69,6 +69,24 @@ export const JETPACK_PRODUCT_DISPLAY_NAMES = {
 	...JETPACK_BACKUP_PRODUCT_DISPLAY_NAMES,
 };
 
+export const JETPACK_BACKUP_PRODUCT_DAILY_TAG_LINE = translate(
```
Isn't `tagline` a single word? If so, let's ditch that underscore in all those constants.

delawski replied:
The dictionary states that the primary form is a "tag line" but a "tagline" as a single word is also accepted. Since it's simpler, I'm going to change it here and in other locations affected by this PR to a single word form 🙂

tyxla replied:
Fair enough that they're both correct. I guess I mean that the single word form is more common around here in the WP context 😉

tyxla commented on client/lib/products-values/index.js:
```diff
@@ -2,11 +2,16 @@
  * External dependencies
  */
 import { assign, difference, get, includes, isEmpty, pick } from 'lodash';
+import moment from 'moment';
```
`moment` is large, and in Calypso we're trying to ditch any direct usages like this in favor of the HoC and hook that we have. See below for a suggestion on how to approach that.

tyxla commented on client/lib/products-values/index.js:
```diff
@@ -359,6 +364,25 @@ export function getDomain( product ) {
+	return JETPACK_PRODUCT_TAG_LINES?.[ product?.productSlug ];
+}
+
+export function isProductExpiring( product ) {
```
I don't think this helper makes a whole lot of sense in this file. It feels it should be defined in the component that consumes it. There is also another reason to move it there: the `moment` usage. 

I suggest that you move this as a method to the component, and wrap that component with the `withLocalizedMoment` HoC if it's a `React.Component`, and with `useLocalizedMoment` hook if it's a functional component.

delawski replied:
Sounds good. I'll move it to the component and use Hoc there.

tyxla commented on client/lib/products-values/index.js:
```diff
@@ -359,6 +364,25 @@ export function getDomain( product ) {
+	return JETPACK_PRODUCT_DISPLAY_NAMES?.[ product?.productSlug ];
+}
+
+export function getJetpackProductTagLine( product ) {
```
Since `tagline` is a single word, we'll also have to rename `TagLine` to `Tagline` here.

tyxla commented on client/lib/products-values/index.js:
```diff
@@ -359,6 +364,25 @@ export function getDomain( product ) {
+	product = formatProduct( product );
+	assertValidProduct( product );
+
+	return JETPACK_PRODUCT_TAG_LINES?.[ product?.productSlug ];
```
Wont the previous `formatProduct` and `assertValidProduct` ensure that we have a valid `product` with a valid `productSlug`, making this `product?.` optional chaining unnecessary?

tyxla commented on client/lib/products-values/index.js:
```diff
@@ -359,6 +364,25 @@ export function getDomain( product ) {
+	product = formatProduct( product );
+	assertValidProduct( product );
+
+	return JETPACK_PRODUCT_DISPLAY_NAMES?.[ product?.productSlug ];
```
Wont the previous `formatProduct` and `assertValidProduct` ensure that we have a valid `product` with a valid `productSlug`, making this `product?.` optional chaining unnecessary?

tyxla commented on client/my-sites/plans/current-plan/purchases-listing.jsx:
```diff
@@ -17,151 +17,208 @@ import {
+
+		if ( currentPlanSlug ) {
+			const planObject = getPlan( currentPlanSlug );
+			return planObject?.getTitle?.();
```
I discourage against using optional chaining for cases like this. If there's no plan object here, then there's something else that's wrong. And if the plan object doesn't have `getTitle` there's something else wrong. We shouldn't be concealing these errors, instead we should be making sure they don't occur (each plan object MUST have a `getTitle()` method).

enejb replied:
:+1:

delawski replied:
Makes sense to me 👍

delawski commented:
@enejb @tyxla I've addressed your feedback in the last 3 commits:

* [Use tagline instead of tag line](https://github.com/Automattic/wp-calypso/pull/37727/commits/8133701e41116b28fa977e911be63f8f9b5e71ef)
* [Get rid of unnecessary optional chaining](https://github.com/Automattic/wp-calypso/pull/37727/commits/7f9189f5bc02e08525f540d928ee7689bb943a03)
* [Move the expiration check to the component and improve the inner logic](https://github.com/Automattic/wp-calypso/pull/37727/commits/229d894fd8dc06b979748396c76e17778ede5920)

Please have another look.

tyxla commented on client/my-sites/plans/current-plan/my-plan-card/index.jsx:
```diff
@@ -29,7 +29,7 @@ const MyPlanCard = ( { action, isError, isPlaceholder, details, plan, tagLine, t
 				<div className="my-plan-card__icon">{ plan && <PlanIcon plan={ plan } /> }</div>
```
Note that I get this when I have a backup product:

![](https://cldup.com/sYn0oJpMr2.png)

I guess it's expected before we land the ProductIcon component and use it instead?

delawski replied:
Well, both yes and no. I'm going to create another 'integration' PR once it's all merged so that we're using the `ProductIcon` here. Since the products are not yet live, there's no way a user would get this error even right now. But yeah - if #37760 is merged first, it'd be the best.

folletto commented:
> Also, I have a question: should we loop in someone from WP.com design? The thing is that this also affects the WP.com plan page. 

In which way a DotCom site is going to be affected? Can you show a before/after that is DotCom specific? Isn't this just for Jetpack sites?

tyxla commented:
> In which way a DotCom site is going to be affected? Can you show a before/after that is DotCom specific? Isn't this just for Jetpack sites?

Well, actually, we changed the layout in a previous PR, and that also affected the WP.com sites. For example, you can see the "My Plan" heading - that was introduced as part of this PR series. Maybe @delawski can give a better list of things that were changed on this page as he worked on them more.

folletto commented:
Ok. Ideally a heading like that should show up only on ours — exactly to not affect DotCom. In general this should be addressed by design before we go to PR, so I'd suggest to ping our designers for review, and highlight if there's any change for DotCom too. They can then assess if it's worth a second review with the other teams or not :)

delawski commented on client/my-sites/plans/current-plan/purchases-listing.jsx:
```diff
@@ -17,151 +17,218 @@ import {
+		purchases: PropTypes.array,
 
 		// From localize() HoC
+		moment: PropTypes.func.isRequired,
```
Thanks for this comment. I wasn't aware of that change.

delawski replied:
This is now fixed.

delawski commented on client/lib/products-values/index.js:
```diff
@@ -359,6 +363,20 @@ export function getDomain( product ) {
 	return product.meta;
 }
 
+export function getJetpackProductDisplayName( product ) {
```
Cool, I'm going to add some test and docblocks to new helpers.

delawski replied:
JSDocs and tests are added now. @tyxla please have a look.

delawski commented on client/lib/products-values/constants.js:
```diff
@@ -69,6 +69,24 @@ export const JETPACK_PRODUCT_DISPLAY_NAMES = {
 	...JETPACK_BACKUP_PRODUCT_DISPLAY_NAMES,
 };
 
+export const JETPACK_BACKUP_PRODUCT_DAILY_TAGLINE = translate(
```
Sure, I can do that 👍

delawski replied:
I'm going to keep it as is for now because we've got more constant names like this (e.g. `JETPACK_BACKUP_PRODUCT_DAILY_DISPLAY_NAME` or `PRODUCT_JETPACK_BACKUP_REALTIME_DESCRIPTION`). I don't want to "pollute" this PR with more changes.

delawski commented:
@folletto @tyxla I can confirm - those changes and the ones from the previous PR (#37642) affect both WPCOM and Jetpack users since they share the same route and page.

If we wanted to revert the page layout for WPCOM users we would have to revive the recently removed `CurrentPlanHeader` component and do some lifting to the `CurrentPlan` component as well. I'd say that code-wise it's a sub-optimal change, but of course if it makes sense UX-wise we can do it.

tyxla commented:
> If we wanted to revert the page layout for WPCOM users we would have to revive the recently removed `CurrentPlanHeader` component and do some lifting to the `CurrentPlan` component as well. I'd say that code-wise it's a sub-optimal change, but of course if it makes sense UX-wise we can do it.

Yep, I agree with Piotr on this one, and I think that it's best for Calypso and WP.com as a product if we keep things consistent and don't add extra maintenance cost.

delawski commented:
@eeeeevon13 @jeffgolenski @crunnells We would like to have your eyes on this update. The issue is discussed here starting with [this comment](https://github.com/Automattic/wp-calypso/pull/37727#issuecomment-556018910).

The merits is that the change of the layout of "My Plan" page affects not only Jetpack users but also WPCOM users. As I mentioned in the [last comment](https://github.com/Automattic/wp-calypso/pull/37727#issuecomment-556037633), reverting this change is sub-optimal code-wise but doable.

Is it okay to leave the "My Plan" page layout as is or not?

Here's a link to the live preview (please compare a Jetpack site with a regular WPCOM site): https://hash-229d894fd8dc06b979748396c76e17778ede5920.calypso.live/plans/my-plan/

Here are some screenshots, too.

**Before**
![Screenshot 2019-11-15 at 18 13 05](https://user-images.githubusercontent.com/478735/69250008-2c61a300-0baf-11ea-9e24-34642c2d58e2.png)

**After**
![Screenshot 2019-11-15 at 18 12 12](https://user-images.githubusercontent.com/478735/69250016-32578400-0baf-11ea-8ac1-d07ab884935a.png)

crunnells commented:
> Is it okay to leave the "My Plan" page layout as is or not?

I think it's probably fine to leave it as-is rather than revert it.
